### DISCUSSION TITLE
Variance of arguments for Generic ABC vs Generic Protocol

### DISCUSSION THREAD

**Discussion by kbehlers** [python/typing#1793] [Q&A] [answered]
I am trying to define a Generic base class, but see different type checking behavior when I inherit from abc.ABC versus Protocol. I am trying to figure out if it is really a difference of variance or if it is just an implementation detail of the type checkers. I've tried against [pyright](https://pyright-play.net/?code=JYWwDg9gTgLgBAQwEYGMBQAzKERxgTzGADsBzOUSWOAcQFNi6pgUAaOABWxghQgBt2AFUJ0AJgBEWMYaIBqCKGmVoU-BAGcNcAEKa6AJToaArvxgAKEWHFSUMAJQAuNHDdwARF9fujp8z5uXh4qBnAAvHDWdApQFh4GHuxIECbEYuF6GobGZo5oBgD6fMQwUAgRUfKK8UUlZQhJcClpGVk5-jJw9eUAborACKXhQlAmdA4qapraemJcEDx8-H55AOrMMEwWC0sCANoGALrOgZ7e7lU2cP3MyPx0nolwJtliFMRwpAxMLHBg3F4Ak8uyBK1y5g2wC2UA8cA0AAtUvx3khHj0ELdBqU3BYoHQqDAAJLEfr8YBiaKxACq2Uml2CykuAAFkCgAHTIDQNewgOgwJFiM5iOgYOAAd02dAsZ0u2X4GFYsvc%2BM6TjgBiVlwccAAtAA%2BOAAOQgjBclwZFwtbihWzwEDgIu5JAQMGApuVQW8ymmWjgAEEdABhVaQqVxNnswNB9j0RjMFCHE7m9yMy5CBHAbQEa4oBF0FAAa20ACtXvAMCQ6MkTPBxQjXRRtGJgBgMEwGPA%2BOBFOJ7Xh85xAcs4HQAI4mYBkzsAfjOjLOrNQnKQ3PKvP5guFoolUpl1rc8sVnrgqry6s1Zx1BuNproKetaYPtsePEdxjdxFd7uIJ4Xvu0INTQaUFllDGAXziUCDjqYDymTecrTcDMszwURunzItS3LOBK0YOBiAgcUB0bINijgiohneWDSnKRB8ReN5EJCRdIy5HkYD5AUICFS4RTFSVoWlE8jy1A8z3MC9yNohAxLca9DRNM0-yQi0X37J1P2-D0D3OFjVHUP0SSxIYYGg8FOkgnZhxghDLRCdNM2zdC8wLYs4DLblcKrdgiWaOhyToXpHjRFAEFeR5oVw6AUGMDV%2BzRD4TNKOd7KZdwlw5dj104zceO3AS9xEgLj10iSYAvOS4AUjUHwtJ9rXUt9NJdN0dIPBqLRPNAgA) and [mypy](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=85a958ad9c0b99be1d53eb14081ef29f).

Comparing `BadProtocolResultWriter` and `ABCResultWriter`, do Generic ABCs just not get typed checked for variance? Or is there something about an ABC that restricts the variance of subclasses?

```
import abc
from typing import Generic, Protocol, TypedDict, TypeVar



class BaseResult(TypedDict):
    """
    Result
    """


R = TypeVar("R", bound=BaseResult)
R_contra = TypeVar("R_contra", bound=BaseResult, contravariant=True)


class BadProtocolResultWriter(Protocol[R]):
    """
    Type variable "R" used in generic protocol "ProtocolResultWriter" should be contravariant  (reportInvalidTypeVarUse)
    main.py:16: error: Invariant type variable "R" used in protocol where contravariant one is expected  [misc]
    """

    @abc.abstractmethod
    def write(
        self,
        result: R,
    ) -> None:
        """
        Write to destination
        """


class ABCResultWriter(abc.ABC, Generic[R]):
    """
    This type checks just fine, but what is different compared to the Protocol equivalent?
    """

    @abc.abstractmethod
    def write(
        self,
        result: R,
    ) -> None:
        """
        Write to destination
        """

class ContraProtocolResultWriter(Protocol[R_contra]):
    """
    This type checks just fine now that C_contra and R_contra are used
    """

    @abc.abstractmethod
    def write(
        self,
        result: R_contra,
    ) -> None:
        """
        Write to destination
        """


class InvariantProtocolResultWriter(Protocol[R]):
    """
    This type checks just fine, I believe because it forces R to be invariant?
    """

    @abc.abstractmethod
    def write(
        self,
        result: R,
    ) -> R:
        """
        Write to destination
        """
  ```

**Comment by erictraut** [comment, marked as answer]
> do Generic ABCs just not get typed checked for variance

Full TypeVar variance consistency is checked only for protocols (as indicated in the [typing spec](https://typing.readthedocs.io/en/latest/spec/protocol.html#generic-protocols)), not for nominal class definitions (including ABCs). Both pyright and mypy do minimal checks for improper usage of covariant and contravariant TypeVars within method parameters and return types, but this isn't the same as full variance consistency checks.

The new TypeVar syntax introduced in PEP 695 eliminates the need to worry about variance consistency because variance is calculated by the type checker based on usage. Pyright has had support for PEP 695 for a long time already, and the next release of mypy will also include support for it. 

Code sample in [pyright playground](https://pyright-play.net/?code=JYWwDg9gTgLgBAQwEYGMBQAzKERxgTzGADsBzOUSWOABWxghQgBsAaOAFUIFMATAEWAoYaNCmYIAzpLgAhKdwBK3SQFdmMABRcwfQcICUALjRwzcAHRXR4qTPm86EBk2bK1GgOpRgMblABtRSM5BXd1GABdTScXFmNTcwABZBQLZEkYKARhEG4YAAsIXkSzXm4MOAB3Hz9NSW5mDHYoFQiQxQM4AFoAPjgAOQhibhNzcctrMQlpOABBWQBhcK9a-yCQ%2BQaVqM1UiwXFhPGU1HSkTOzc-KKS8fLKmt9uesbmuFaPGA6uvsHh0alCZWCw2GYyRbDLIIWKMFg7bzPQLBULbNoaaKw1zHZL7DLQ66FYpAh7VNavJotdHfOCdHr9IYjMYTMwgsF2OAASWIADcED4EMQYFj4dTEX5kZswtTMfQ4cwcWZTml8VcYHkiXdzKSnnUGpSPtSfvTacyWSCgA)

```python
import abc
from typing import Protocol, TypedDict

class BaseResult(TypedDict):
    ...

class BadProtocolResultWriter[R: BaseResult](Protocol):
    @abc.abstractmethod
    def write(self, result: R) -> None:
        ...

class ABCResultWriter[R: BaseResult](abc.ABC):
    @abc.abstractmethod
    def write(self, result: R) -> None:
        ...

class ContraProtocolResultWriter[R: BaseResult](Protocol):
    @abc.abstractmethod
    def write(self, result: R) -> None:
        ...

class InvariantProtocolResultWriter[R: BaseResult](Protocol):
    @abc.abstractmethod
    def write(self, result: R) -> R:
        ...
```

If you're using pyright or pylance, you can hover over the TypeVar declaration to see the computed variance.

![image](https://github.com/python/typing/assets/7040122/2dc411c0-d1c8-4137-a520-6361efd50417)

![image](https://github.com/python/typing/assets/7040122/5f55fc57-f7cf-4419-91fe-9573327cb5dd)

**Reply by kbehlers** [reply]
Thank you, that answers my question! This was very enlightening, and I appreciate the tips about pyright's PEP 695 support and hovering over the TypeVar declaration!
